### PR TITLE
Interchain Accounts / lupercalia

### PR DESCRIPTION
**Pull request by faddat**
This PR will bring ICA to Juno, as well as:

* wasmd v0.24.0
* ibc-go
* an upgrade handler folder to better organize code
* go 1.18
* tighter linting

This PR should be applied after a governance "Software Upgrade" proposal to juno passes.  

The governance "Software upgrade" proposal should not be made until this code has been merged to Juno's main branch.

### CODE REVIEW

**Comment by faddat** [comment]
Awesome @nghuyenthevinh2000!

Ok, now we just add the ica stuff which is here:

https://github.com/cosmos/ibc-go/blob/main/testing/simapp/app.go

**Comment by faddat** [comment]
we should also use wasmd from the same commit (nearly) as Osmosis:

**Comment by faddat** [comment]
I am making a cosmos-sdk fork with @pomvn's fast iavl patch only, so we don't depend on osmosis-sdk

**Comment by faddat** [comment]
https://github.com/notional-labs/cosmos-sdk/tree/v0.45.1-junofast

**Comment by giansalex** [comment]
we can keep this pull request only for ibc-go v3 and ICA, and move iavl and other forks to other pull requests.
I also heard that iavl fork throws app_hash errors in osmosis, or does it have another cause?

File: go.mod [hunk 1]
```diff
@@ -123,7 +124,9 @@ require (
 )
 
 replace (
-	github.com/cosmos/iavl => github.com/cosmos/iavl v0.17.3
+	github.com/cosmos/cosmos-sdk => github.com/notional-labs/cosmos-sdk v0.45.1-allfast
+	github.com/cosmos/iavl => github.com/osmosis-labs/iavl v0.17.3-osmo-v5.0.20220321022426-8371754f3299
+	github.com/cosmwasm/wasmd => github.com/notional-labs/wasmd v0.21.1-0.20220319082325-607293cd0bfe
```
Comment (the-frey) [inline go.mod:129]: Why are we moving off mainline wasmd?

The whole point of lupercalia was to move _to_ mainline wasmd and collect security patches along the way.

Just trying to understand reasoning here :)

Comment (giansalex) [reply go.mod:129]: these forks should be removed from this pull request

Comment (faddat) [reply go.mod:129]: ah ok so the reasoning is interchain accounts

no fork no ica

wdyt

**Comment by faddat** [comment]
no one really knows about those apphash errors, we can take fast iavl out

**Comment by nguyenvuong1122000** [comment]
@faddat, i think we should remove iavl from wasmd

**Comment by faddat** [comment]
yes you're right and thanks very much vuong. 

It was kind of all or nothing with fast iavl.

**Comment by faddat** [comment]
@vuong177 @nghuyenthevinh2000 

Hey guys, this needs to be reworked.  I'll click around a little, and see how much is needed.

**Comment by faddat** [comment]
Hey, I am going to close this in favor of #200 

While writing that, it became apparent that it might as well be this.

**Pull request closed without merging**